File: frontend/src/obs/routines.ts
import {
  ApplyStreamInfo,
  ApplyStreamInfoForPlan,
  EndStreamSession,
  GetActiveStreamSession,
  GetPlannedStreams,
  GetPlanSessions,
  GetRoutines,
  GetStreamdeckMultiActions,
  PressHotkey,
} from '../../wailsjs/go/main/App'
import {main} from '../../wailsjs/go/models'
import {pushEpisodeText} from '../lib/smartSources'

/**
 * Routine execution. A routine's normalized steps (authored in Jax, or parsed
 * from a Stream Deck Multi Action by the backend — see streamdeck.go) are
 * replayed over the app's obs-websocket connection. The Stream Deck software
 * has no API to press a button remotely, so "managed with Streamdeck" means
 * Jax re-reads the Multi Action from the Stream Deck's profile files at run
 * time and performs the steps itself; steps Jax cannot perform (third-party
 * plugins like Philips Hue) are skipped and reported.
 */

/** The ServicesProvider's obsRequest shape. */
export type ObsRequest = <T = Record<string, unknown>>(
  type: string,
  data?: Record<string, unknown>,
) => Promise<T>

/** IDs/triggers of the two built-in routines (mirrors routines.go). */
export const START_ROUTINE = 'start-stream'
export const END_ROUTINE = 'end-stream'

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms))

/** A human label for one step, used in lists and previews. */
export function describeStep(step: main.RoutineStep): string {
  switch (step.kind) {
    case 'obs-scene':
      return `Switch scene to “${step.scene}”`
    case 'obs-source':
      return `Toggle source “${step.source}” in “${step.scene}”`
    case 'obs-mute':
      return step.mode === 'mute'
        ? `Mute “${step.source}”`
        : step.mode === 'unmute'
          ? `Unmute “${step.source}”`
          : `Toggle mute on “${step.source}”`
    case 'obs-stream':
      return step.mode === 'start'
        ? 'Start the stream'
        : step.mode === 'stop'
          ? 'Stop the stream'
          : 'Toggle the stream'
    case 'obs-record':
      return step.mode === 'start'
        ? 'Start recording'
        : step.mode === 'stop'
          ? 'Stop recording'
          : 'Toggle recording'
    case 'delay':
      return `Wait ${((step.delayMs ?? 0) / 1000).toLocaleString()}s`
    case 'hotkey':
      return `Press ${step.description || 'a hotkey'}`
    case 'update-smart-sources':
      return 'Update episode info (mapped text sources)'
    case 'apply-stream-info':
      return 'Apply stream info (title, description, tags & category)'
    case 'streamdeck':
      return `Stream Deck: “${step.description || 'Multi Action'}”`
    default:
      return step.description || 'Unsupported step'
  }
}

/**
 * A routine's two phases, ready to run: `before` runs ahead of the built-in
 * routines' stream transition, `after` once the stream has started/stopped.
 * Custom routines have no transition and only use `before`. Warnings collect
 * anything that could not be resolved (e.g. a deleted Multi Action).
 */
export interface RoutinePhases {
  before: main.RoutineStep[]
  after: main.RoutineStep[]
  warnings: string[]
}

/**
 * Resolve the steps a routine will run right now. Stream Deck Multi Action
 * steps are expanded into the Multi Action's own steps, re-read from the
 * deck's profile files on every run so edits made in the Stream Deck app are
 * always picked up; a missing Multi Action skips that step with a warning.
 */
export async function resolveRoutinePhases(
  routine: main.Routine,
): Promise<RoutinePhases> {
  const warnings: string[] = []
  const all = [...(routine.steps ?? []), ...(routine.afterSteps ?? [])]

  let actions: main.StreamdeckMultiAction[] = []
  if (all.some((s) => s.kind === 'streamdeck')) {
    try {
      actions = (await GetStreamdeckMultiActions()) ?? []
    } catch (err) {
      warnings.push(
        err instanceof Error && err.message
          ? err.message
          : 'Could not read the Stream Deck profiles.',
      )
    }
  }

  const expand = (steps: main.RoutineStep[]): main.RoutineStep[] =>
    (steps ?? []).flatMap((s) => {
      if (s.kind !== 'streamdeck') return [s]
      const match = actions.find((a) => a.id === s.streamdeckActionId)
      if (!match) {
        warnings.push(
          `The Stream Deck Multi Action “${
            s.description || 'Multi Action'
          }” was not found — its steps were skipped.`,
        )
        return []
      }
      return match.steps ?? []
    })

  return {
    before: expand(routine.steps ?? []),
    after: expand(routine.afterSteps ?? []),
    warnings,
  }
}

/** Perform one step. Unsupported steps report themselves via the return. */
async function runStep(
  step: main.RoutineStep,
  obsRequest: ObsRequest,
  opts: RunRoutineOptions = {},
): Promise<string | null> {
  switch (step.kind) {
    case 'delay':
      await sleep(step.delayMs ?? 0)
      return null

    case 'hotkey':
      // A Stream Deck Hotkey button: the backend synthesizes the keystroke
      // system-wide, exactly as pressing the deck button would.
      await PressHotkey(
        step.vkey ?? 0,
        Boolean(step.ctrl),
        Boolean(step.shift),
        Boolean(step.alt),
        Boolean(step.win),
      )
      return null

    case 'obs-scene': {
      // The Stream Deck OBS plugin's "preview" target only differs from
      // "program" while studio mode is on; mirror that behaviour.
      let studioMode = false
      if (step.target === 'preview') {
        try {
          const r = await obsRequest<{studioModeEnabled: boolean}>(
            'GetStudioModeEnabled',
          )
          studioMode = Boolean(r.studioModeEnabled)
        } catch {
          studioMode = false
        }
      }
      await obsRequest(
        studioMode ? 'SetCurrentPreviewScene' : 'SetCurrentProgramScene',
        {sceneName: step.scene},
      )
      return null
    }

    case 'obs-source': {
      let itemId = step.sceneItemId ?? 0
      if (!itemId) {
        const r = await obsRequest<{sceneItemId: number}>('GetSceneItemId', {
          sceneName: step.scene,
          sourceName: step.source,
        })
        itemId = r.sceneItemId
      }
      const {sceneItemEnabled} = await obsRequest<{sceneItemEnabled: boolean}>(
        'GetSceneItemEnabled',
        {sceneName: step.scene, sceneItemId: itemId},
      )
      await obsRequest('SetSceneItemEnabled', {
        sceneName: step.scene,
        sceneItemId: itemId,
        sceneItemEnabled: !sceneItemEnabled,
      })
      return null
    }

    case 'obs-mute':
      if (step.mode === 'mute' || step.mode === 'unmute') {
        await obsRequest('SetInputMute', {
          inputName: step.source,
          inputMuted: step.mode === 'mute',
        })
      } else {
        await obsRequest('ToggleInputMute', {inputName: step.source})
      }
      return null

    case 'obs-stream':
      await obsRequest(
        step.mode === 'start'
          ? 'StartStream'
          : step.mode === 'stop'
            ? 'StopStream'
            : 'ToggleStream',
      )
      return null

    case 'obs-record':
      await obsRequest(
        step.mode === 'start'
          ? 'StartRecord'
          : step.mode === 'stop'
            ? 'StopRecord'
            : 'ToggleRecord',
      )
      return null

    case 'update-smart-sources':
      return updateEpisodeSmartSources(obsRequest, Boolean(opts.test))

    case 'apply-stream-info': {
      // Push the on-air planned stream's info to its channels: Twitch's
      // channel info, and YouTube's live broadcast (or, off the air, its
      // upcoming default/scheduled broadcast — what YouTube Studio edits
      // offline). The backend no-ops when no planned stream is on the air —
      // except in a test, which pushes the ready-to-go-live plan's info for
      // real; both platforms simply apply it to the next stream.
      if (opts.test) {
        const session = await GetActiveStreamSession()
        if (!session.active || !session.planId) {
          const plan = await nextPlannedStream()
          if (!plan) {
            return 'Test: no planned stream is ready to go live, so no stream info was applied.'
          }
          const warnings = (await ApplyStreamInfoForPlan(plan.id)) ?? []
          return [
            `Test: applied “${plan.title}” stream info to its channels.`,
            ...warnings,
          ].join(' · ')
        }
      }
      const warnings = (await ApplyStreamInfo()) ?? []
      return warnings.length > 0 ? warnings.join(' · ') : null
    }

    default:
      return `Skipped (Stream Deck only): ${step.description || step.kind}`
  }
}

/**
 * The "Update episode info" step: write the on-air planned stream's episode
 * title and "Episode N" directly into the OBS text sources mapped on the
 * series' edit page — plain text, no tokens or templates involved.
 *
 * Going live with a plan opens its stream session before the Start routine
 * runs, so the step sees the new episode. When no planned stream is on the
 * air the step is a silent no-op — except in a test, which rehearses with the
 * planned stream that is ready to go live, exactly what the real go-live
 * would show.
 */
async function updateEpisodeSmartSources(
  obsRequest: ObsRequest,
  test: boolean,
): Promise<string | null> {
  const session = await GetActiveStreamSession()
  if (session.active) {
    if (!session.title && session.episode <= 0) return null
    await pushEpisodeText(obsRequest, session.title, session.episode)
    return null
  }
  if (!test) return null

  const plan = await nextPlannedStream()
  if (!plan) {
    return 'Test: no planned stream is ready to go live, so the episode text sources were left as-is.'
  }
  if (!plan.title && plan.episodeNumber <= 0) return null
  const wrote = await pushEpisodeText(obsRequest, plan.title, plan.episodeNumber)
  if (!wrote) {
    return 'Test: no episode text sources are mapped — choose them on the series’ edit page.'
  }
  return `Test: episode sources show “${plan.title}”${
    plan.episodeNumber > 0 ? ` (Episode ${plan.episodeNumber})` : ''
  } from the upcoming plan.`
}

/**
 * The planned stream a test should rehearse with: the newest plan that has
 * not been broadcast yet — the same card "Go Live with Planned Stream" puts
 * on top with its Go Live button ready.
 */
async function nextPlannedStream(): Promise<main.PlannedStream | null> {
  try {
    const [plans, sessions] = await Promise.all([
      GetPlannedStreams(),
      GetPlanSessions(),
    ])
    const broadcast = new Set((sessions ?? []).map((s) => s.planId))
    return (plans ?? []).find((p) => !broadcast.has(p.id)) ?? null
  } catch {
    return null
  }
}

/**
 * Run a list of steps in order. A failing step does not abort the routine;
 * every problem is collected and returned as human-readable warnings.
 */
export async function runSteps(
  steps: main.RoutineStep[],
  obsRequest: ObsRequest,
  opts: RunRoutineOptions = {},
): Promise<string[]> {
  const warnings: string[] = []
  for (const step of steps) {
    try {
      const skipped = await runStep(step, obsRequest, opts)
      if (skipped) warnings.push(skipped)
    } catch (err) {
      warnings.push(
        `${describeStep(step)} failed${
          err instanceof Error && err.message ? `: ${err.message}` : ''
        }`,
      )
    }
  }
  return warnings
}

export interface RunRoutineOptions {
  /**
   * Test mode: rehearse the routine without touching the broadcast. Every
   * step runs as usual except stream start/stop/toggle steps (and the
   * built-ins' implied stream transition), which are skipped and reported,
   * and the End routine leaves the stream session open.
   */
  test?: boolean
}

/**
 * Run one routine: its before-steps, then — for the built-in routines — the
 * stream transition, then its after-steps. A broken or missing phase only
 * produces warnings; the built-ins' stream transition itself is guaranteed
 * (unless testing, which never starts or stops the stream).
 */
export async function runRoutine(
  routine: main.Routine,
  obsRequest: ObsRequest,
  opts: RunRoutineOptions = {},
): Promise<string[]> {
  const warnings: string[] = []
  let phases: RoutinePhases = {before: [], after: [], warnings: []}
  try {
    phases = await resolveRoutinePhases(routine)
    warnings.push(...phases.warnings)
  } catch (err) {
    warnings.push(
      err instanceof Error && err.message
        ? err.message
        : 'The routine could not be loaded.',
    )
  }
  if (opts.test) {
    const holdStream = (steps: main.RoutineStep[]) =>
      steps.filter((s) => {
        if (s.kind !== 'obs-stream') return true
        warnings.push(`Test: skipped “${describeStep(s)}”.`)
        return false
      })
    phases = {
      ...phases,
      before: holdStream(phases.before),
      after: holdStream(phases.after),
    }
  }
  warnings.push(...(await runSteps(phases.before, obsRequest, opts)))
  if (opts.test) {
    if (routine.trigger === START_ROUTINE) {
      warnings.push('Test: the stream was not started.')
    } else if (routine.trigger === END_ROUTINE) {
      warnings.push('Test: the stream was not stopped.')
    }
  } else {
    warnings.push(
      ...(await impliedStreamTransition(routine.trigger, phases.before, obsRequest)),
    )
  }
  warnings.push(...(await runSteps(phases.after, obsRequest, opts)))
  // Ending the broadcast closes the open stream session (the record a
  // planned stream's chat and transcript are attached to). A test is not an
  // ending: whatever is on the air stays attached to its session.
  if (routine.trigger === END_ROUTINE && !opts.test) {
    await EndStreamSession().catch(() => {})
  }
  return warnings
}

/**
 * Run the built-in routine behind the app's Go live / Stop stream buttons.
 */
export async function runStreamRoutine(
  trigger: typeof START_ROUTINE | typeof END_ROUTINE,
  obsRequest: ObsRequest,
): Promise<string[]> {
  let routine: main.Routine | undefined
  try {
    routine = ((await GetRoutines()) ?? []).find((r) => r.trigger === trigger)
  } catch {
    // Fall through: the stream transition must happen regardless.
  }
  if (!routine) return impliedStreamTransition(trigger, [], obsRequest)
  return runRoutine(routine, obsRequest)
}

/** Start/stop the stream for a built-in routine when no step already has. */
async function impliedStreamTransition(
  trigger: string,
  steps: main.RoutineStep[],
  obsRequest: ObsRequest,
): Promise<string[]> {
  if (trigger !== START_ROUTINE && trigger !== END_ROUTINE) return []
  if (steps.some((s) => s.kind === 'obs-stream')) return []
  try {
    await obsRequest(trigger === START_ROUTINE ? 'StartStream' : 'StopStream')
  } catch (err) {
    // Already in the requested state (e.g. a Run of "Start Stream" while
    // live) is not worth surfacing; anything else is.
    const message = err instanceof Error ? err.message : ''
    if (!/already/i.test(message)) {
      return [
        trigger === START_ROUTINE
          ? `Could not start the stream${message ? `: ${message}` : ''}`
          : `Could not stop the stream${message ? `: ${message}` : ''}`,
      ]
    }
  }
  return []
}

File: routines.go
package main

import (
	"fmt"
	"log"
	"strings"
	"time"
)

// ---------------------------------------------------------------------------
// Routines
//
// A Routine is a sequence of broadcast actions the app can run — switching
// scenes, waiting, muting inputs, starting/stopping the stream. Two built-in
// routines ("Start Stream" / "End Stream") are tied to the app's go-live and
// stop-stream buttons; additional routines can be added and run manually from
// the OBS Studio → Routines tab.
//
// A routine is managed either by Jax (its steps are defined in the app) or by
// a Stream Deck Multi Action (the steps are read from the Stream Deck's own
// profile files and replayed by Jax — see streamdeck.go). Stored as a single
// JSON blob in the settings table.
// ---------------------------------------------------------------------------

// The retired "manage with Streamdeck" mode, recognised by migrateRoutine.
const routineManagerStreamdeck = "streamdeck"

// Reserved IDs (and triggers) of the two built-in routines.
const (
	routineStartStream = "start-stream"
	routineEndStream   = "end-stream"
)

// RoutineStep is one normalized action in a routine. Kind decides which of
// the other fields apply. The same shape is used for steps authored in Jax
// and steps parsed out of a Stream Deck Multi Action.
type RoutineStep struct {
	// Kind: "obs-scene" | "obs-source" | "obs-mute" | "obs-stream" |
	// "obs-record" | "delay" | "streamdeck" | "unsupported".
	Kind string `json:"kind"`
	// Scene name (obs-scene: the scene to switch to; obs-source: the scene
	// holding the item).
	Scene string `json:"scene,omitempty"`
	// Target of a scene switch: "program" (default) or "preview".
	Target string `json:"target,omitempty"`
	// Source name (obs-source: the scene item; obs-mute: the input).
	Source string `json:"source,omitempty"`
	// SceneItemID of an obs-source step ("0" = resolve by Source name).
	SceneItemID int `json:"sceneItemId,omitempty"`
	// Mode: obs-mute "toggle"|"mute"|"unmute"; obs-stream/obs-record
	// "toggle"|"start"|"stop".
	Mode string `json:"mode,omitempty"`
	// DelayMs of a delay step.
	DelayMs int `json:"delayMs,omitempty"`
	// StreamdeckActionID references a Stream Deck Multi Action whose steps
	// are replayed in place of this step at run time (kind "streamdeck");
	// Description carries its title for display when the deck is unavailable.
	StreamdeckActionID string `json:"streamdeckActionId,omitempty"`
	// Description of an unsupported step (e.g. "Philips Hue: Color") — these
	// steps still run on the Stream Deck itself but are skipped when Jax
	// replays the routine — or a streamdeck step's Multi Action title.
	Description string `json:"description,omitempty"`
}

// Routine is one runnable routine.
type Routine struct {
	ID   string `json:"id"`
	Name string `json:"name"`
	// Trigger ties the routine to an app action: "start-stream" and
	// "end-stream" run with the corresponding stream buttons; "" is manual.
	Trigger string `json:"trigger"`
	// BuiltIn routines are pinned: they always exist and cannot be deleted.
	BuiltIn bool `json:"builtIn"`
	// Deprecated: routines used to be managed either "with Jax" (authored
	// steps) or "with Streamdeck" (bound to Multi Actions). A Multi Action is
	// now just a step kind; these fields only remain so previously stored
	// routines decode, and normalizeRoutines migrates them into Steps.
	Manager                 string `json:"manager,omitempty"`
	StreamdeckActionID      string `json:"streamdeckActionId,omitempty"`
	StreamdeckTitle         string `json:"streamdeckTitle,omitempty"`
	StreamdeckAfterActionID string `json:"streamdeckAfterActionId,omitempty"`
	StreamdeckAfterTitle    string `json:"streamdeckAfterTitle,omitempty"`
	// The built-in routines run in two phases around their stream transition:
	// Steps runs before it, AfterSteps once it has happened. Custom routines
	// have no transition and only use Steps.
	Steps      []RoutineStep `json:"steps"`
	AfterSteps []RoutineStep `json:"afterSteps"`
	CreatedAt  string        `json:"createdAt"`
}

// builtinRoutines returns the two pinned routines in display order.
func builtinRoutines() []Routine {
	return []Routine{
		{
			ID:         routineStartStream,
			Name:       "Start Stream",
			Trigger:    routineStartStream,
			BuiltIn:    true,
			Steps:      []RoutineStep{},
			AfterSteps: []RoutineStep{},
		},
		{
			ID:         routineEndStream,
			Name:       "End Stream",
			Trigger:    routineEndStream,
			BuiltIn:    true,
			Steps:      []RoutineStep{},
			AfterSteps: []RoutineStep{},
		},
	}
}

// normalizeRoutines overlays the stored routines on the built-in set, so the
// two pinned routines always exist (keeping any stored configuration) and
// custom routines follow in their stored order. Routines stored in the
// retired two-manager shape are migrated on the way out.
func normalizeRoutines(stored []Routine) []Routine {
	out := builtinRoutines()
	custom := []Routine{}
	for _, r := range stored {
		r = migrateRoutine(r)
		switch r.ID {
		case routineStartStream:
			r.BuiltIn, r.Trigger = true, routineStartStream
			out[0] = r
		case routineEndStream:
			r.BuiltIn, r.Trigger = true, routineEndStream
			out[1] = r
		default:
			r.BuiltIn = false
			custom = append(custom, r)
		}
	}
	return append(out, custom...)
}

// migrateRoutine converts the retired two-manager shape into the unified step
// list: a routine that was "managed with Streamdeck" becomes one whose phases
// each hold a single streamdeck step referencing its Multi Action. Authored
// steps stored while that manager was active never ran, so they are dropped
// to preserve behaviour.
func migrateRoutine(r Routine) Routine {
	if r.Manager == routineManagerStreamdeck {
		r.Steps = []RoutineStep{}
		r.AfterSteps = []RoutineStep{}
		if r.StreamdeckActionID != "" {
			r.Steps = append(r.Steps, RoutineStep{
				Kind:               "streamdeck",
				StreamdeckActionID: r.StreamdeckActionID,
				Description:        r.StreamdeckTitle,
			})
		}
		if r.StreamdeckAfterActionID != "" {
			r.AfterSteps = append(r.AfterSteps, RoutineStep{
				Kind:               "streamdeck",
				StreamdeckActionID: r.StreamdeckAfterActionID,
				Description:        r.StreamdeckAfterTitle,
			})
		}
	}
	r.Manager = ""
	r.StreamdeckActionID, r.StreamdeckTitle = "", ""
	r.StreamdeckAfterActionID, r.StreamdeckAfterTitle = "", ""
	return r
}

// GetRoutines returns every routine, the two built-ins first. Never nil.
func (a *App) GetRoutines() []Routine {
	if a.store == nil {
		return builtinRoutines()
	}
	var stored []Routine
	if _, err := a.store.getJSON(keyRoutines, &stored); err != nil {
		log.Printf("jax: GetRoutines: %v", err)
	}
	return normalizeRoutines(stored)
}

// SaveRoutine upserts a routine (matched by ID), assigning an ID and creation
// time on first save, and returns the stored value.
func (a *App) SaveRoutine(routine Routine) (Routine, error) {
	if a.store == nil {
		return routine, fmt.Errorf("storage unavailable")
	}
	if strings.TrimSpace(routine.Name) == "" {
		return routine, fmt.Errorf("give the routine a name")
	}
	routine = migrateRoutine(routine)
	if routine.Steps == nil {
		routine.Steps = []RoutineStep{}
	}
	if routine.AfterSteps == nil {
		routine.AfterSteps = []RoutineStep{}
	}
	for _, s := range append(append([]RoutineStep{}, routine.Steps...), routine.AfterSteps...) {
		if s.Kind == "streamdeck" && s.StreamdeckActionID == "" {
			return routine, fmt.Errorf("choose a Multi Action for each Stream Deck step")
		}
	}

	all := a.GetRoutines()
	if routine.ID == "" {
		routine.ID = fmt.Sprintf("routine_%d", time.Now().UnixNano())
	}
	if routine.CreatedAt == "" {
		routine.CreatedAt = time.Now().UTC().Format(time.RFC3339)
	}

	replaced := false
	for i, r := range all {
		if r.ID == routine.ID {
			// The pinned identity of a built-in survives whatever the form sent.
			routine.BuiltIn, routine.Trigger = r.BuiltIn, r.Trigger
			all[i] = routine
			replaced = true
			break
		}
	}
	if !replaced {
		all = append(all, routine)
	}

	if err := a.store.setJSON(keyRoutines, all); err != nil {
		return routine, err
	}
	return routine, nil
}

// DeleteRoutine removes a custom routine by ID; the built-ins are pinned.
func (a *App) DeleteRoutine(id string) error {
	if a.store == nil {
		return fmt.Errorf("storage unavailable")
	}
	if id == routineStartStream || id == routineEndStream {
		return fmt.Errorf("the built-in stream routines cannot be deleted")
	}
	all := a.GetRoutines()
	out := make([]Routine, 0, len(all))
	for _, r := range all {
		if r.ID != id {
			out = append(out, r)
		}
	}
	return a.store.setJSON(keyRoutines, out)
}
